File: training/src/compiler/training/base/optimizers/LAMB.cpp
// Copyright (C) 2022. Huawei Technologies Co., Ltd. All rights reserved.

// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include "LAMB.h"
#include <algorithm>
#include <iostream>
#include <stdexcept>

namespace
{
constexpr raul::dtype lowerBoundary = 0.0_dt;
constexpr raul::dtype upperBoundary = 1.0_dt;

constexpr raul::dtype weightNormLowerBoundary = 0.0_dt;
constexpr raul::dtype weightNormUpperBoundary = 10.0_dt;
}

namespace raul::optimizers
{

LAMB::LAMB(const dtype lr, const dtype beta1, const dtype beta2, const dtype epsilon, const dtype weightDecay, const bool adam)
    : mLearningRate(lr)
    , mBeta1(beta1)
    , mBeta2(beta2)
    , mEpsilon(epsilon)
    , mWeightDecay(weightDecay)
    , mAdam(adam)
{
    if (lr < lowerBoundary)
    {
        THROW_NONAME("LAMB", "reset lr>" + Conversions::toString(lowerBoundary) + " (current lr=" + Conversions::toString(lr) + ")");
    }
    if (epsilon < lowerBoundary)
    {
        THROW_NONAME("LAMB", "reset epsilon>" + Conversions::toString(lowerBoundary) + " (current epsilon=" + Conversions::toString(epsilon) + ")");
    }

    if (beta1 < lowerBoundary || beta1 >= upperBoundary)
    {
        THROW_NONAME("LAMB", "reset beta1 from [" + Conversions::toString(lowerBoundary) + ", " + Conversions::toString(upperBoundary) + ") (current beta1=" + Conversions::toString(beta1) + ")");
    }
    if (beta2 < lowerBoundary || beta2 >= upperBoundary)
    {
        THROW_NONAME("LAMB", "reset beta2 from [" + Conversions::toString(lowerBoundary) + ", " + Conversions::toString(upperBoundary) + ") (current beta2=" + Conversions::toString(beta1) + ")");
    }
}

void LAMB::optimize(MemoryManager& memory_manager, Tensor& param, const Tensor& grad)
{
    if (param.size() != grad.size())
    {
        THROW_NONAME("LAMB", "parameters and gradients must have the same size (" + Conversions::toString(param.size()) + " != " + Conversions::toString(grad.size()) + ")");
    }

    if (!memory_manager.tensorExists(Name("LAMB") / param.getName() / "m"))
    {
        // Exponential moving average of gradient values
        memory_manager.createTensor(Name("LAMB") / param.getName() / "m", 1, param.size(), 1, 1, 0.0_dt);
        // Exponential moving average of squared gradient values
        memory_manager.createTensor(Name("LAMB") / param.getName() / "v", 1, param.size(), 1, 1, 0.0_dt);
        // Temporal tensor to keep actual parameter updates
        memory_manager.createTensor(Name("LAMB") / param.getName() / "adam_steps", 1, param.size(), 1, 1, 0.0_dt);
    }

    Tensor& m = memory_manager[Name("LAMB") / param.getName() / "m"];
    Tensor& v = memory_manager[Name("LAMB") / param.getName() / "v"];
    Tensor& adamSteps = memory_manager[Name("LAMB") / param.getName() / "adam_steps"];

    // Calculate weight norm and adam norm
    dtype weightNorm = 0.0_dt;
    dtype adamNorm = 0.0_dt;
#if defined(_OPENMP)
#pragma omp parallel for reduction(+ : weightNorm, adamNorm)
#endif
    for (size_t i = 0; i < grad.size(); ++i)
    {
        // Decay the first and second moment running average coefficient
        m[i] = m[i] * mBeta1 + (1.0_dt - mBeta1) * grad[i];
        v[i] = v[i] * mBeta2 + (1.0_dt - mBeta2) * grad[i] * grad[i];
        // Adam steps
        adamSteps[i] = m[i] / (std::sqrt(v[i]) + mEpsilon) + mWeightDecay * param[i];
        // Weight and adam norms
        weightNorm += param[i] * param[i];
        adamNorm += adamSteps[i] * adamSteps[i];
    }
    weightNorm = std::clamp(std::sqrt(weightNorm), weightNormLowerBoundary, weightNormUpperBoundary);
    adamNorm = std::sqrt(adamNorm);

    // Update parameter
    OPENBLAS_CONST dtype sa = -mLearningRate * ((mAdam || weightNorm == 0.0_dt || adamNorm == 0.0_dt) ? 1.0_dt : weightNorm / adamNorm);
    OPENBLAS_CONST dtype* sx = &(adamSteps[0]);
    size_t incx = 1U;
    dtype* sy = &(param[0]);
    size_t incy = 1U;
    size_t xOffset = 0U;
    size_t yOffset = 0U;
    Common::axpy(param.size(), sa, sx, incx, sy, incy, xOffset, yOffset);
}

void LAMB::optimize(MemoryManagerFP16& memory_manager, TensorFP16& param, const TensorFP16& grad)
{
    if (param.size() != grad.size())
    {
        THROW_NONAME("LAMB", "parameters and gradients must have the same size (" + Conversions::toString(param.size()) + " != " + Conversions::toString(grad.size()) + ")");
    }

    if (!memory_manager.tensorExists(Name("LAMB") / param.getName() / "m"))
    {
        // Exponential moving average of gradient values
        memory_manager.createTensor(Name("LAMB") / param.getName() / "m", 1, param.size(), 1, 1, 0.0_hf);
        // Exponential moving average of squared gradient values
        memory_manager.createTensor(Name("LAMB") / param.getName() / "v", 1, param.size(), 1, 1, 0.0_hf);
        // Temporal tensor to keep actual parameter updates
        memory_manager.createTensor(Name("LAMB") / param.getName() / "adam_steps", 1, param.size(), 1, 1, 0.0_hf);
    }

    TensorFP16& m = memory_manager[Name("LAMB") / param.getName() / "m"];
    TensorFP16& v = memory_manager[Name("LAMB") / param.getName() / "v"];
    TensorFP16& adamSteps = memory_manager[Name("LAMB") / param.getName() / "adam_steps"];

    // Calculate weight norm and adam norm
    dtype weightNorm = 0.0_dt;
    dtype adamNorm = 0.0_dt;
#if defined(_OPENMP)
#pragma omp parallel for reduction(+ : weightNorm, adamNorm)
#endif
    for (size_t i = 0; i < grad.size(); ++i)
    {
        // Decay the first and second moment running average coefficient
        const auto mTmp = TODTYPE(m[i]) * mBeta1 + (1.0_dt - mBeta1) * TODTYPE(grad[i]);
        m[i] = TOHTYPE(mTmp);
        const auto vTmp = TODTYPE(v[i]) * mBeta2 + (1.0_dt - mBeta2) * TODTYPE(grad[i]) * TODTYPE(grad[i]);
        v[i] = TOHTYPE(vTmp);
        // Adam steps
        const auto adamStepTmp = mTmp / (std::sqrt(vTmp) + mEpsilon) + mWeightDecay * TODTYPE(param[i]);
        adamSteps[i] = TOHTYPE(adamStepTmp);
        // Weight and adam norms
        weightNorm += TODTYPE(param[i]) * TODTYPE(param[i]);
        adamNorm += adamStepTmp * adamStepTmp;
    }
    weightNorm = std::clamp(std::sqrt(weightNorm), weightNormLowerBoundary, weightNormUpperBoundary);
    adamNorm = std::sqrt(adamNorm);

    // Update parameter
    OPENBLAS_CONST dtype sa = -mLearningRate * ((mAdam || weightNorm == 0.0_dt || adamNorm == 0.0_dt) ? 1.0_dt : weightNorm / adamNorm);
    OPENBLAS_CONST half* sx = &(adamSteps[0]);
    size_t incx = 1U;
    half* sy = &(param[0]);
    size_t incy = 1U;
    size_t xOffset = 0U;
    size_t yOffset = 0U;
    Common::axpy(param.size(), sa, sx, incx, sy, incy, xOffset, yOffset);
}

std::ostream& LAMB::as_ostream(std::ostream& out) const
{
    out << "LAMB(lr=" << std::scientific << mLearningRate << ", beta1=" << mBeta1 << ", beta2=" << mBeta2;
    out << ", epsilon=" << mEpsilon << ", weight decay=" << mWeightDecay << ", adam: " << (mAdam ? "true" : "false") << ")";
    return out;
}

} // raul::optimizers

File: training/src/tests/tests/python/LAMB.py
# Copyright (C) 2022. Huawei Technologies Co., Ltd. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

"""Lamb optimizer."""
import collections
import math
import torch
from torch.optim import Optimizer


class Lamb(Optimizer):
    r"""Implements Lamb algorithm.
    It has been proposed in `Large Batch Optimization for Deep Learning: Training BERT in 76 minutes`_.
    Arguments:
        params (iterable): iterable of parameters to optimize or dicts defining
            parameter groups
        lr (float, optional): learning rate (default: 1e-3)
        betas (Tuple[float, float], optional): coefficients used for computing
            running averages of gradient and its square (default: (0.9, 0.999))
        eps (float, optional): term added to the denominator to improve
            numerical stability (default: 1e-8)
        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
        adam (bool, optional): always use trust ratio = 1, which turns this into
            Adam. Useful for comparison purposes.
    .. _Large Batch Optimization for Deep Learning: Training BERT in 76 minutes:
    """

    def __init__(
        self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-6, weight_decay=0, adam=False
    ):
        if not 0.0 <= lr:
            raise ValueError("Invalid learning rate: {}".format(lr))
        if not 0.0 <= eps:
            raise ValueError("Invalid epsilon value: {}".format(eps))
        if not 0.0 <= betas[0] < 1.0:
            raise ValueError("Invalid beta parameter at index 0: {}".format(betas[0]))
        if not 0.0 <= betas[1] < 1.0:
            raise ValueError("Invalid beta parameter at index 1: {}".format(betas[1]))

        defaults = dict(lr=lr, betas=betas, eps=eps, weight_decay=weight_decay)
        self.adam = adam
        super(Lamb, self).__init__(params, defaults)

    def step(self, closure=None):
        """Performs a single optimization step.
        Arguments:
            closure (callable, optional): A closure that reevaluates the model
                and returns the loss.
        """
        loss = None
        if closure is not None:
            loss = closure()
        for group in self.param_groups:
            for p in group["params"]:
                if p.grad is None:
                    continue
                grad = p.grad.data
                if grad.is_sparse:
                    raise RuntimeError(
                        "Lamb does not support sparse gradients, consider SparseAdam instad."
                    )

                state = self.state[p]

                # State initialization
                if len(state) == 0:
                    state["step"] = 0
                    # Exponential moving average of gradient values
                    state["exp_avg"] = torch.zeros_like(p.data)
                    # Exponential moving average of squared gradient values
                    state["exp_avg_sq"] = torch.zeros_like(p.data)

                exp_avg, exp_avg_sq = state["exp_avg"], state["exp_avg_sq"]
                beta1, beta2 = group["betas"]
                state["step"] += 1

                # Decay the first and second moment running average coefficient
                # m_t
                exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
                # v_t
                exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)

                # Paper v3 does not use debiasing.
                # bias_correction1 = 1 - beta1 ** state['step']
                # bias_correction2 = 1 - beta2 ** state['step']
                # Apply bias to lr to avoid broadcast.
                step_size = group[
                    "lr"
                ]  # * math.sqrt(bias_correction2) / bias_correction1
                weight_norm = p.data.pow(2).sum().sqrt().clamp(0, 10)
                adam_step = exp_avg / exp_avg_sq.sqrt().add(group["eps"])
                if group["weight_decay"] != 0:
                    adam_step.add_(p.data, alpha=group["weight_decay"])

                adam_norm = adam_step.pow(2).sum().sqrt()
                if weight_norm == 0 or adam_norm == 0:
                    trust_ratio = 1
                else:
                    trust_ratio = weight_norm / adam_norm
                state["weight_norm"] = weight_norm
                state["adam_norm"] = adam_norm
                state["trust_ratio"] = trust_ratio
                if self.adam:
                    trust_ratio = 1
                p.data.add_(adam_step, alpha=-step_size * trust_ratio)

        return loss


# Double step unit (no weight decay)
torch.manual_seed(0)
torch.set_printoptions(precision=8)
param = torch.rand(1, 2, 3, 4)
param.grad = torch.rand(1, 2, 3, 4)
print("Parameter(before update): ", param)
print("Gradient: ", param.grad)
optimizer = Lamb([param], lr=0.1, betas=(0.5, 0.8), eps=1)
# First step
optimizer.step()
print("Parameter(after first step): ", param)
# Second step
optimizer.step()
print("Parameter(after second step): ", param)

# Double step unit (adam style)
param = torch.rand(1, 2, 3, 4)
param.grad = torch.rand(1, 2, 3, 4)
print("Parameter(before update): ", param)
print("Gradient: ", param.grad)
optimizer = Lamb([param], lr=0.1, betas=(0.5, 0.8), eps=1, weight_decay=0.1, adam=True)
# First step
optimizer.step()
print("Parameter(after first step): ", param)
# Second step
optimizer.step()
print("Parameter(after second step): ", param)

# Double step unit
param = torch.rand(1, 2, 3, 4)
param.grad = torch.rand(1, 2, 3, 4)
print("Parameter(before update): ", param)
print("Gradient: ", param.grad)
optimizer = Lamb([param], lr=0.1, betas=(0.5, 0.8), eps=1, weight_decay=0.1, adam=False)
# First step
optimizer.step()
print("Parameter(after first step): ", param)
# Second step
optimizer.step()
print("Parameter(after second step): ", param)
